src/lib.rs

```diff
@@ -0,0 +1,450 @@
+extern crate sxd_document;
+
+use std::str;
+use std::string::String;
+use std::collections::HashMap;
+
+use sxd_document::{
+    parser,
+    Package,
+};
+use sxd_document::dom::{
+    Document,
+    Root,
+    ChildOfRoot,
+    Element,
+    ChildOfElement,
+};
+
+#[derive(Debug)]
+pub struct Program {
+    pub groups: Vec<StatementBody>
+}
+
+#[derive(PartialEq, Debug)]
+pub struct StatementBody {
+    pub blocks: Vec<Block>
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Block {
+    pub block_type: String,
+    pub id: String,
+    pub fields: Option<HashMap<String, FieldValue>>,
+    pub statements: Option<HashMap<String, StatementBody>>,
+}
+
+#[derive(PartialEq, Debug)]
+pub enum FieldValue {
+    SimpleField(String),
+    ExpressionField(Block),
+}
+
+impl Program {
+    pub fn new() -> Self {
+        Self {
+            groups: Vec::new()
+        }
+    }
+}
+
+impl StatementBody {
+    pub fn new() -> Self {
+        Self {
+            blocks: Vec::new()
+        }
+    }
+}
+
+impl Block {
+    pub fn new() -> Self {
+        Self {
+            block_type: "".to_string(),
+            id: "".to_string(),
+            fields: None,
+            statements: None
+        }
+    }
+
+    pub fn add_field(&mut self, name: String, value: FieldValue) {
+        match self.fields {
+            Some(ref mut f) => {
+                f.insert(name.to_owned(), value);
+            },
+            None => {
+                let mut new_map: HashMap<String, FieldValue> = HashMap::new();
+                new_map.insert(name.to_owned(), value);
+                self.fields = Some(new_map);
+            }
+        }
+    }
+
+    pub fn get_field(&self, name: &str) -> Option<&FieldValue> {
+        match self.fields {
+            Some(ref fields) => {
+                fields.get(name)
+            },
+            None => None
+        }
+    }
+
+    pub fn add_statement_body(&mut self, name: String, body: StatementBody) {
+        match self.statements {
+            Some(ref mut s) => {
+                s.insert(name.to_owned(), body);
+            },
+            None => {
+                let mut new_map: HashMap<String, StatementBody> = HashMap::new();
+                new_map.insert(name.to_owned(), body);
+                self.statements = Some(new_map);
+            }
+        };
+    }
+
+    pub fn get_statement(&self, name: &str) -> Option<&StatementBody> {
```

Comment:
This is a place where the borrow checker seemed to force me to return a reference to a `StatementBody`, but I didn't have that problem in other places. Is there a way around that?

In general, is it more preferred for a public method/function to return an `Option<MyEntity>` or an `Option<&MyEntity>`?

Reply:
In this case, the data is owned by the `statements` hashmap. Since you do not want to *move* the `StatementBody` out of the hashmap, you want to simply return a reference to it.

If something is in a collection, and you want to keep it there (e.g. not call `collection.remove(element)` then in general you want to use a reference.

Reply:
Thanks for the tip, that makes more sense than before! I think in this case it actually makes sense for every method to return a borrow, since these data structures are meant to be an immutable intermediate representation, and application developers would want to make their own domain-specific structures from these.